### DISCUSSION TITLE
Binary Data Float64Array Positions Clarification

### DISCUSSION THREAD

**Discussion by mcost45** [visgl/deck.gl#6139] [Q&A] [answered]
Hi, I've been reading up about 64 vs 32 bit precision. Based on the binary data documentation examples I've seen, I've been using Float64Arrays to hold my position values, with standard layers.

Does this mean 64 bit projection will be used automatically, with the increased performance cost? And if I instead use Float32Arrays for my positions, 32 bit projection would be used instead? Or is 32 projection always used by default?

**Comment by Pessimistress** [comment, marked as answer]
deck.gl handles positions with pseudo-64-bit precision (emulated with 2 32-bit floats) which is less accurate than a true double but sufficient in most use cases.

On the shader side (where projection is performed), it is extremely efficient (one extra subtraction, which does not have any meaningful cost), and is always used. If you provide a `Float32Array` instead of `Float64Array`, the low part of the two floats is simply treated as `vec3(0.0)`.

However, when you update the positions with binary input, handling `Float64Array` is more expensive because we 

- allocate a temporary `Float32Array` that is twice as long
- split each double value into two floats using `Math.fround` and stack them into the temp array
- upload the temp array to the GPU

So comparing to supplying `Float32Array`, there is extra work on the CPU, twice as much CPU>GPU data transfer, as well as memory usage.

Note that the above still happens very fast, so the perf difference is not usually noticeable. When people choose `Float32Array` over `Float64Array`, the much more significant benefit is to only send/receive half as much data from the server. If you don't care for high accuracy, for example the source data is in lower fidelity, or the application does not require high zoom (<12) then `Float32Array` is recommended.

**Reply by mcost45** [reply]
Ahh I see, thank you for all the details!
